mis/solver/greedymapping.py

```diff
@@ -0,0 +1,261 @@
+    """
+
+    def __init__(
+        self,
+        instance: MISInstance,
+        lattice: Lattice,
+        previously_generated_subgraphs: List[Dict[int, int]],
+        seed: int = 0,
+    ) -> None:
+        """
+        Initializes the GreedyMapping algorithm for mapping a graph onto a lattice.
+
+        Args:
+            instance: The MIS problem instance containing the logical graph.
+            lattice: The lattice structure defining the physical layout.
+            previous_mappings: List of previous mappings (for scoring reuse).
+            seed: Random seed for reproducibility.
+        """
+        self.graph: nx.Graph = instance.graph.copy()
+        self.lattice_instance: Lattice = lattice
+        self.lattice: nx.Graph = nx.convert_node_labels_to_integers(self.lattice_instance.lattice)
+        self.previously_generated_subgraphs = previously_generated_subgraphs
+        random.seed(seed)
+
+    def generate(self,
+                 starting_node: int,
+        remove_invalid_placement_nodes: bool = True,
+        rank_nodes: bool = True,
+    ) -> Dict[int, int]:
+        """
+        Generates a subgraph by mapping the input graph onto the lattice using a greedy approach.
+
+        Args:
+            starting_node: The initial graph node to start mapping.
+            remove_invalid_placement_nodes: Whether to remove invalid placements.
+            rank_nodes: Whether to rank nodes using the scoring heuristic.
+
+        Returns:
+            dict: A dictionary representing the graph-to-lattice mapping.
+        """
+        unmapping: Dict[int, int] = {}
+        mapping: Dict[int, int] = {}
+        unexpanded_nodes: Set[int] = set()
+
+        current_lattice_node: int = self._initialize(
+            starting_node, mapping, unmapping, unexpanded_nodes
+        )
+        current_node: int = starting_node
+
+        while unexpanded_nodes:
+            unexpanded_nodes.remove(current_node)
+
+            lattice_neighbors = list(self.lattice.neighbors(current_lattice_node))
+            free_lattice_neighbors = [
+                neighbor for neighbor in lattice_neighbors if neighbor not in unmapping
+            ]
+
+            neighbors = list(self.graph.neighbors(current_node))
+
+            self._extend_mapping(
+                considered_nodes=neighbors,
+                unexpanded_nodes=unexpanded_nodes,
+                free_lattice_neighbors=free_lattice_neighbors,
+                mapping=mapping,
+                unmapping=unmapping,
+                remove_invalid_placement_nodes=remove_invalid_placement_nodes,
+                rank_nodes=rank_nodes,
+            )
+
+            if unexpanded_nodes:
+                current_node = next(iter(unexpanded_nodes))
+                current_lattice_node = mapping[current_node]
+
+        if not self._validate(mapping, unmapping):
+            raise Exception("Invalid mapping!")
```

Comment:
Do we need custom exceptions for graph errors handling ?

Reply:
Since we're writing a library, I agree that we probably want custom exceptions. Our users may want to catch them, after all.

Reply:
What I meant was "let's declare a subclass of `Exception` or perhaps of `ValueError` and use it here.